### ISSUE TITLE
PostgreSQL MetaImpl based DataTypes are missing length of array element type

### ISSUE THREAD

**Issue by rbuehlma** [jOOQ/jOOQ#11307]
### Expected behavior

The metamodel of an array with an element type of defined length should contain the length on the element type.

### Actual behavior

The element type does not contain a length.

### Steps to reproduce the problem

This code illustrates to issue:

```
@Test
    public void testCreateTable() throws Exception {
        Connection conn = getSourceDbConnection();
        conn.createStatement().execute("create table test_table (col5 varchar(10)[] not null)");
        DSLContext dsl = DSL.using(conn);
        org.jooq.Table<?> testTable = dsl.meta().getSchemas("test-source").get(0).getTable("test_table");
        Field field = testTable.field(0);
        DataType dataType = field.getDataType(dsl.configuration());
        DataType elementType = dataType.getArrayComponentDataType();
        assertThat(elementType.length(), is(10));
    }
```

### Versions

- jOOQ: jOOQ-3.15.0-SNAPSHOT-20210126004003.zip
- Java: 11
- Database (include vendor): Postgresql 10.7
- OS: Linux
- JDBC Driver (include name if inofficial driver): org.postgresql (42.2.5)

**Comment by lukaseder** [comment]
Thanks a lot for your report. I can reproduce this, the reported component data type is `VARCHAR`, not `VARCHAR(10)`. Will look into this right away.

**Comment by lukaseder** [comment]
This isn't PostgreSQL specific, it also fails on HSQLDB

**Comment by lukaseder** [comment]
This can be solved in two ways:

- Fix it in `MetaImpl` only, where the information from JDBC's `DatabaseMetaData` is turned into a `DataType`
- Change the behaviour of `DataType.hasLength` (and precision, scale, etc.) as well as that of `DataType.length()` (and precision, scale, etc.) to operate on the array component type, in case a type is an array.

I tend to think the real issue is the second one and that should be fixed. Currently, setting the length on an array type has no effect. Turning an `VARCHAR(10)[]` type into a `VARCHAR(20)[]` type is quite laborious through jOOQ API:

```java
DataType<String[]> t1 = VARCHAR(10).getArrayType();
DataType<String[]> t2 = t1.getArrayComponentType().length(20).getArrayType();
```

It could be as simple as `VARCHAR(10).getArrayType().length(20)`. What other thing could it mean for a user to call `length(20)` on a `VARCHAR(10)[]` type?

However, client code may have hard-wired assumptions that `DataType.hasLength() == true` means the type is a `VARCHAR` or `VARBINARY` or similar, and never an array. The question is, can this be seen as a bug or a missing feature. I tend to see it as a bug.

Note that both PostgreSQL and Oracle support fixed sized arrays (see https://github.com/jOOQ/jOOQ/issues/5932). It will be a good idea not to recycle the term "length" in that case, but instead introduce either `arrayLength` or `size` (like JDK's `Collection::size`)

**Comment by lukaseder** [comment]
While `DataType<T[]>.length(x)` can be reasonably retrofitted to be short for `DataType<T[]>.getArrayComponentType().length(20).getArrayType()`, this isn't necessarily true for all the other properties.

These property setters could be delegated to the component type on the grounds of the fact that they have no meaning on the array type itself:

- Precision
- Scale
- Length
- Collation
- CharacterSet

These properties make no sense at all in the context of arrays, so the discussion is irrelevant to them:

- Identity

But these properties (and possibly future ones like them) can live independently on the array type and on the component type:

- Nullability
- DefaultValue

E.g. it's possible to think of an array whose elements cannot be null, but the array itself can be. Or vice versa, an array cannot be null, but it's elements can be. In kotlin notation: `Array<T?>` and `Array<T>?` are both distinct types.

From a forward compatibility and consistency perspective, it would be unwise to cleverly propagate some properties to the component type, but not others. So, the fix will be done purely in `MetaImpl` for now.

**Comment by lukaseder** [comment]
The motivation behind the above attempts is to achieve greater code re-use. `MetaImpl` isn't really the proper place to apply length/precision/scale to an array or non-array type. Perhaps, this logic could be implemented in ` org.jooq.impl.DefaultDataType.getDataType(SQLDialect, String, int, int)` instead

**Comment by lukaseder** [comment]
The fix won't work yet for arrays of UDTs, because we don't have the full UDT type information yet in `MetaImpl` derived types.

**Comment by lukaseder** [comment]
H2 array types are currently not affected by this change, because H2 does not yet have typed arrays: http://h2database.com/html/datatypes.html#array_type

We'll adapt our implementations once their typed arrays are ready in one of the next releases.

**Comment by lukaseder** [comment]
Unlike the PostgreSQL JDBC driver, HSQLDB's `DatabaseMetaData::getColumns` implementation doesn't provide array element type length/precision/scale. See e.g.

```
+---------+-----------+----------+------------+---------+-----------------+-----------+-------------+--------------+--------------+--------+-------+----------+-------------+----------------+-----------------+----------------+-----------+-------------+------------+-----------+----------------+----------------+
|TABLE_CAT|TABLE_SCHEM|TABLE_NAME|COLUMN_NAME |DATA_TYPE|TYPE_NAME        |COLUMN_SIZE|BUFFER_LENGTH|DECIMAL_DIGITS|NUM_PREC_RADIX|NULLABLE|REMARKS|COLUMN_DEF|SQL_DATA_TYPE|SQL_DATETIME_SUB|CHAR_OCTET_LENGTH|ORDINAL_POSITION|IS_NULLABLE|SCOPE_CATALOG|SCOPE_SCHEMA|SCOPE_TABLE|SOURCE_DATA_TYPE|IS_AUTOINCREMENT|
+---------+-----------+----------+------------+---------+-----------------+-----------+-------------+--------------+--------------+--------+-------+----------+-------------+----------------+-----------------+----------------+-----------+-------------+------------+-----------+----------------+----------------+
|PUBLIC   |PUBLIC     |T_ARRAYS  |ID          |        4|INTEGER          |         32|{null}       |             0|             2|       0|{null} |{null}    |            4|          {null}|                0|               1|NO         |{null}       |{null}      |{null}     |{null}          |NO              |
|PUBLIC   |PUBLIC     |T_ARRAYS  |STRING_ARRAY|     2003|VARCHAR(20) ARRAY|          0|{null}       |        {null}|        {null}|       1|{null} |{null}    |           50|          {null}|                0|               2|YES        |{null}       |{null}      |{null}     |{null}          |NO              |
|PUBLIC   |PUBLIC     |T_ARRAYS  |NUMBER_ARRAY|     2003|INTEGER ARRAY    |          0|{null}       |        {null}|        {null}|       1|{null} |{null}    |           50|          {null}|                0|               3|YES        |{null}       |{null}      |{null}     |{null}          |NO              |
|PUBLIC   |PUBLIC     |T_ARRAYS  |DATE_ARRAY  |     2003|DATE ARRAY       |          0|{null}       |        {null}|        {null}|       1|{null} |{null}    |           50|          {null}|                0|               4|YES        |{null}       |{null}      |{null}     |{null}          |NO              |
+---------+-----------+----------+------------+---------+-----------------+-----------+-------------+--------------+--------------+--------+-------+----------+-------------+----------------+-----------------+----------------+-----------+-------------+------------+-----------+----------------+----------------+
```

We should probably parse the `TYPE_NAME` information, but I feel like a prior refactoring of some of `DefaultDataType` internal workings would be helpful here.

I'll make this fix about PostgreSQL only, which is a lower hanging fruit.

**Comment by lukaseder** [comment]
HSQLDB specific issue: https://github.com/jOOQ/jOOQ/issues/11311

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.6 (#11312)
